v2/pkg/genruntime/extensions/prereconciliation_checker.go

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package extensions
+
+import (
+	"context"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/conditions"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+)
+
+// PreReconciliationChecker is implemented by resources that want to do extra checks before proceeding with
+// a full ARM reconcile.
+type PreReconciliationChecker interface {
+	// PreReconcileCheck does a pre-reconcile check to see if the resource is in a state that can be reconciled.
+	// ARM resources should implement this to avoid reconciliation attempts that cannot possibly succeed.
+	// Returns ProceedWithReconcile if the reconciliation should go ahead.
+	// Returns BlockReconcile and a human-readable reason if the reconciliation should be skipped.
+	// ctx is the current operation context.
+	// obj is the resource about to be reconciled. The resource's State will be freshly updated.
+	// kubeClient allows access to the cluster for any required queries.
```

Comment:
minor: missing owner documentation here -- although see above question about if it'd be better to just pass `ResourceResolver` instead